File: tools/softirqslower.py
#!/usr/bin/env python
# @lint-avoid-python-3-compatibility-imports
#
# softirqlat  Trace slow soft IRQ (interrupt).
#             For Linux, uses BCC, eBPF.
#
# USAGE: softirqslower [-h] [-c CPU] [min_us]
#
# Copyright (c) 2025 Chenyue Zhou.
# Licensed under the Apache License, Version 2.0 (the "License")
#
# 08-Jul-2025   Chenyue Zhou     Created this.

from __future__ import print_function
from bcc import BPF
from time import sleep, strftime
import argparse
import sys

# arguments
examples = """examples:
    ./softirqslower        # trace softirq latency higher than 10000 us (default)
    ./softirqslower 100000 # trace softirq latency higher than 100000 us
    ./softirqslower -c 1   # trace softirq latency on CPU 1 only
"""

parser = argparse.ArgumentParser(
        description="Trace slow soft IRQ (interrupt).",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=examples)
parser.add_argument("min_us", nargs="?", default="10000",
        help="minimum softirq latency to trace, in us (default 10000)")
parser.add_argument("-c", "--cpu", type=int, help="trace this CPU only")
parser.add_argument("--ebpf", action="store_true", help=argparse.SUPPRESS)
args = parser.parse_args()

bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>
#include <linux/interrupt.h>

enum {
    SOFTIRQ_RAISE,
    SOFTIRQ_ENTRY,
    SOFTIRQ_EXIT,
    SOFTIRQ_MAX_STAGE,
};

struct counter {
    u64 ts;
    u8 not_first;
};

struct data {
    u32 reserved;
    u32 stage;
    u32 vec;
    u32 cpu;
    u64 delta_ns;
    char task[TASK_COMM_LEN];
};

BPF_PERCPU_ARRAY(raise, struct counter, NR_SOFTIRQS);
BPF_PERCPU_ARRAY(entry, struct counter, NR_SOFTIRQS);
BPF_PERF_OUTPUT(softirq_events);

static __always_inline void event_collect(void *ctx, u32 stage, u32 vec,
    u32 cpu, u64 delta_ns)
{
    struct data dt = {
        .stage = stage,
        .vec = vec,
        .cpu = cpu,
        .delta_ns = delta_ns,
    };
    bpf_get_current_comm(&dt.task, sizeof(dt.task));
    softirq_events.perf_submit(ctx, &dt, sizeof(dt));
}

RAW_TRACEPOINT_PROBE(softirq_raise)
{
    u32 cpu = bpf_get_smp_processor_id();
    u32 vec = (u32)(ctx->args[0]);

    FILTER_CPU

    struct counter *data = raise.lookup(&vec);
    if (!data)
        return 0;

    if (data->ts) {
        // TODO record event
        return 0;
    }

    data->not_first = 1;
    data->ts = bpf_ktime_get_ns();

    return 0;
}

RAW_TRACEPOINT_PROBE(softirq_entry)
{
    u32 cpu = bpf_get_smp_processor_id();
    u32 vec = (u32)(ctx->args[0]);

    FILTER_CPU

    struct counter *data = raise.lookup(&vec);
    if (!data)
        return 0;

    if ((data->not_first) && !(data->ts)) {
        // TODO record miss event
        return 0;
    }

    u64 cur_ts = bpf_ktime_get_ns();
    u64 delta_ns = cur_ts - data->ts;
    data->ts = 0;

    if (DELTA_FILTER) {
        event_collect(ctx, SOFTIRQ_ENTRY, vec, cpu, delta_ns);
    }
    data = entry.lookup(&vec);
    if (!data)
        return 0;

    data->not_first = 1;
    data->ts = cur_ts;

    return 0;
}

RAW_TRACEPOINT_PROBE(softirq_exit)
{
    u32 cpu = bpf_get_smp_processor_id();
    u32 vec = (u32)(ctx->args[0]);

    FILTER_CPU

    struct counter *data = entry.lookup(&vec);
    if (!data)
        return 0;

    if ((data->not_first) && !(data->ts)) {
        // TODO record miss event
        return 0;
    }

    u64 cur_ts = bpf_ktime_get_ns();
    u64 delta_ns = cur_ts - data->ts;
    data->ts = 0;

    if (DELTA_FILTER) {
        event_collect(ctx, SOFTIRQ_EXIT, vec, cpu, delta_ns);
    }

    return 0;
}
"""

def vec_to_name(vec):
    return ["hi", "timer", "net_tx", "net_rx", "block", "irq_poll", "tasklet",
            "sched", "hrtimer", "rcu"][vec]

def stage_to_name(stage):
    return ["raise softirq", "irq(hard) to softirq", "softirq runtime"][stage]

def print_event(cpu, data, size):
    event = b["softirq_events"].event(data)
    print("%-8s %-20s %-8s %-14d %-6d %-6s" % (strftime("%H:%M:%S"),
            stage_to_name(event.stage), vec_to_name(event.vec),
            event.delta_ns / 1000, event.cpu, event.task.decode("utf-8",
                "replace")))

if __name__ == "__main__":
    if args.cpu is not None:
        bpf_text = bpf_text.replace("FILTER_CPU",
                'if (cpu != %d) { return 0; }' % int(args.cpu))
    else:
        bpf_text = bpf_text.replace("FILTER_CPU", "")

    bpf_text = bpf_text.replace("DELTA_FILTER", "delta_ns >= %d" % \
            (int(args.min_us) * 1000))

    if args.ebpf:
        print(bpf_text)
        exit()

    b = BPF(text=bpf_text)
    b["softirq_events"].open_perf_buffer(print_event, page_cnt=64)

    print("Tracing softirq latency higher than %d us... Hit Ctrl-C to end." % \
            int(args.min_us))

    print("%-8s %-20s %-8s %-14s %-6s %-6s" % ("TIME", "STAGE", "SOFTIRQ",
        "LAT(us)", "CPU", "COMM"))
    while (1):
        try:
            b.perf_buffer_poll()
        except KeyboardInterrupt:
            exit()

File: libbpf-tools/softirqslower.c
// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
// Copyright (c) 2026 Ism Hong
//
// Based on softirqslower(8) from BCC by Chenyue Zhou.
// libbpf/CO-RE version.
// 27-May-2026   Ported to libbpf.
#include <argp.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <bpf/libbpf.h>
#include <bpf/bpf.h>
#include "softirqslower.h"
#include "softirqslower.skel.h"
#include "trace_helpers.h"

#define ARRAY_SIZE(x) (sizeof(x) / sizeof(*(x)))

static volatile sig_atomic_t exiting = 0;

struct env {
	__u64 min_us;
	int   targ_cpu;
	bool  verbose;
} env = {
	.min_us   = 10000,
	.targ_cpu = -1,
};

const char *argp_program_version = "softirqslower 0.1";
const char *argp_program_bug_address =
	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
const char argp_program_doc[] =
"Trace slow soft IRQ (interrupt).\n"
"\n"
"USAGE: softirqslower [--help] [-c CPU] [min_us]\n"
"\n"
"EXAMPLES:\n"
"    softirqslower        # trace softirq latency higher than 10000 us (default)\n"
"    softirqslower 100000 # trace softirq latency higher than 100000 us\n"
"    softirqslower -c 1   # trace softirq latency on CPU 1 only\n";

static const struct argp_option opts[] = {
	{ "cpu",     'c', "CPU",  0, "Trace this CPU only", 0 },
	{ "verbose", 'v', NULL,   0, "Verbose debug output", 0 },
	{ NULL,      'h', NULL,   OPTION_HIDDEN, "Show the full help", 0 },
	{},
};

static error_t parse_arg(int key, char *arg, struct argp_state *state)
{
	static int pos_args;
	long long val;

	switch (key) {
	case 'h':
		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
		break;
	case 'v':
		env.verbose = true;
		break;
	case 'c':
		errno = 0;
		val = strtol(arg, NULL, 10);
		if (errno || val < 0) {
			fprintf(stderr, "invalid cpu: %s\n", arg);
			argp_usage(state);
		}
		env.targ_cpu = (int)val;
		break;
	case ARGP_KEY_ARG:
		if (pos_args++) {
			fprintf(stderr,
				"Unrecognized positional argument: %s\n", arg);
			argp_usage(state);
		}
		errno = 0;
		val = strtoll(arg, NULL, 10);
		if (errno || val <= 0) {
			fprintf(stderr,
				"Invalid min latency (in us): %s\n", arg);
			argp_usage(state);
		}
		env.min_us = (__u64)val;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static int libbpf_print_fn(enum libbpf_print_level level,
		const char *format, va_list args)
{
	if (level == LIBBPF_DEBUG && !env.verbose)
		return 0;
	return vfprintf(stderr, format, args);
}

static void sig_int(int signo)
{
	exiting = 1;
}

/* Softirq vector names – same order as the kernel enum */
static const char *vec_names[] = {
	[0] = "hi",
	[1] = "timer",
	[2] = "net_tx",
	[3] = "net_rx",
	[4] = "block",
	[5] = "irq_poll",
	[6] = "tasklet",
	[7] = "sched",
	[8] = "hrtimer",
	[9] = "rcu",
};

static const char *stage_names[] = {
	[SOFTIRQ_RAISE] = "irq(hard) to softirq",
	[SOFTIRQ_ENTRY] = "softirq runtime",
};

static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
{
	struct event e;
	char ts[32];

	if (data_sz < sizeof(e)) {
		printf("Error: packet too small\n");
		return;
	}
	memcpy(&e, data, sizeof(e));

	str_timestamp("%H:%M:%S", ts, sizeof(ts));

	printf("%-8s %-20s %-8s %-14llu %-6u %-16s\n",
			ts,
			e.stage < ARRAY_SIZE(stage_names) ? stage_names[e.stage] : "unknown",
			e.vec  < NR_SOFTIRQS ? vec_names[e.vec] : "unknown",
			e.delta_us,
			e.cpu,
			e.task);
}

static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
{
	printf("Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
}

int main(int argc, char **argv)
{
	static const struct argp argp = {
		.options = opts,
		.parser  = parse_arg,
		.doc     = argp_program_doc,
	};
	struct perf_buffer       *pb  = NULL;
	struct softirqslower_bpf *obj;
	int err;

	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (err)
		return err;

	libbpf_set_print(libbpf_print_fn);

	obj = softirqslower_bpf__open();
	if (!obj) {
		fprintf(stderr, "failed to open BPF object\n");
		return 1;
	}

	/* Configure rodata constants before loading */
	obj->rodata->min_us   = env.min_us;
	obj->rodata->targ_cpu = env.targ_cpu;

	/* Choose between tp_btf (preferred) and raw_tp (fallback) */
	if (probe_tp_btf("softirq_raise")) {
		/* Kernel supports BTF tracepoints – disable raw_tp variants */
		bpf_program__set_autoload(obj->progs.softirq_raise_raw, false);
		bpf_program__set_autoload(obj->progs.softirq_entry_raw, false);
		bpf_program__set_autoload(obj->progs.softirq_exit_raw,  false);
	} else {
		/* Fall back to raw tracepoints – disable tp_btf variants */
		bpf_program__set_autoload(obj->progs.softirq_raise_btf, false);
		bpf_program__set_autoload(obj->progs.softirq_entry_btf, false);
		bpf_program__set_autoload(obj->progs.softirq_exit_btf,  false);
	}

	err = softirqslower_bpf__load(obj);
	if (err) {
		fprintf(stderr, "failed to load BPF object: %d\n", err);
		goto cleanup;
	}

	err = softirqslower_bpf__attach(obj);
	if (err) {
		fprintf(stderr, "failed to attach BPF programs\n");
		goto cleanup;
	}

	printf("Tracing softirq latency higher than %llu us... "
			"Hit Ctrl-C to end.\n", env.min_us);
	printf("%-8s %-20s %-8s %-14s %-6s %-16s\n",
			"TIME", "STAGE", "SOFTIRQ", "LAT(us)", "CPU", "COMM");

	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), 64,
			handle_event, handle_lost_events, NULL, NULL);
	if (!pb) {
		err = -errno;
		fprintf(stderr, "failed to open perf buffer: %d\n", err);
		goto cleanup;
	}

	if (signal(SIGINT, sig_int) == SIG_ERR) {
		fprintf(stderr, "can't set signal handler: %s\n",
			strerror(errno));
		err = 1;
		goto cleanup;
	}

	while (!exiting) {
		err = perf_buffer__poll(pb, 100);
		if (err < 0 && err != -EINTR) {
			fprintf(stderr, "error polling perf buffer: %s\n",
				strerror(-err));
			goto cleanup;
		}
		/* reset err to return 0 if exiting */
		err = 0;
	}

cleanup:
	perf_buffer__free(pb);
	softirqslower_bpf__destroy(obj);

	return err != 0;
}
